cadc-conesearch/src/main/java/org/opencadc/conesearch/TAPQueryGenerator.java

```diff
@@ -0,0 +1,236 @@
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * Class to produce an ADQL query from a set of Parameters
+ */
+public class TAPQueryGenerator extends CommonParamValidator {
+    private static final int DEF_MAXREC = 1000;
+    private static final int MAX_MAXREC = Integer.MAX_VALUE;
+    private static final int MIN_VERB_VALUE = 1;
+    private static final int MID_VERB_VALUE = 2;
+    private static final int MAX_VERB_VALUE = 3;
+    private static final Logger LOGGER = LogManager.getLogger(TAPQueryGenerator.class);
+
+    protected final String catalog;
+    protected final String lowVerbositySelectList;
+    protected final String midVerbositySelectList;
+    protected final String highVerbositySelectList;
+
+    public TAPQueryGenerator(final String catalog, final String lowVerbositySelectList,
+                             final String midVerbositySelectList, final String highVerbositySelectList) {
+        this.catalog = catalog;
+        this.lowVerbositySelectList = lowVerbositySelectList;
+        this.midVerbositySelectList = midVerbositySelectList;
+        this.highVerbositySelectList = highVerbositySelectList;
+    }
+
+    /**
+     * Map with the REQUEST, LANG, and QUERY parameters.
+     *
+     * @param positionColumnName The name of the positional column to compare against the cone's position,
+     *                           preferably spatially indexed.
+     * @return map of parameter names and values
+     */
+    public Map<String, Object> getParameterMap(final String positionColumnName,
+                                               final Map<String, List<String>> parameters) {
+        final Map<String, Object> queryParameterMap = new HashMap<>();
+
+        // Obtain and, if necessary, provide a default RESPONSEFORMAT.
+        final String requestedResponseFormat = getFirstParameter(ParameterNames.RESPONSEFORMAT.name(), parameters,
+                                                                 VOTableWriter.CONTENT_TYPE);
+        queryParameterMap.put(ParameterNames.RESPONSEFORMAT.name(), requestedResponseFormat);
+
+        // Obtain and validate the VERB (verbosity) output.
+        final String requestedOutputVerbosity = getFirstParameter(ParameterNames.VERB.name(), parameters,
+                                                                  Integer.toString(MID_VERB_VALUE));
+        final NumberParameterValidator outputVerbosityValidator =
+                new NumberParameterValidator(false, TAPQueryGenerator.MIN_VERB_VALUE,
+                                             TAPQueryGenerator.MAX_VERB_VALUE, TAPQueryGenerator.MID_VERB_VALUE);
+        final int outputVerbosity = outputVerbosityValidator.validate(requestedOutputVerbosity);
+
+        // Obtain and validate the MAXREC value with a default if necessary.
+        final String requestedMaxRecords = getFirstParameter(ParameterNames.MAXREC.name(), parameters,
```

Comment:
MAXREC should be added to the CommonParamValidator, eg
```
public Integer getMaxRec(Map<>, Integer defaultValue, Integer maxValue)
```
where both integer args could be null (server-side choice). There is a MaxRecValidator in cadc-dali that is used in TAP, but it currently requires a Job so isn't usable here, but the code in it's validate() method is easily adapted.

Reply:
This is done.